implementations/utils/xpk.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities to run workloads with xpk (https://github.com/google/xpk)."""
+
+import uuid
+from absl import logging
+from airflow.decorators import task
+from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
+from kubernetes import client, config
+
+
+@task
+def generate_workload_id(benchmark_id: str) -> str:
+  """Generate a workload ID."""
+  short_id = str(uuid.uuid4())[:8]
+  return f"{benchmark_id}-{short_id}"
+
+
+def run_workload(
+    task_id: str,
+    project_id: str,
+    zone: str,
+    cluster_name: str,
+    benchmark_id: str,
+    workload_id: str,
+    docker_image: str,
+    accelerator_type: str,
+    run_cmds: str,
+    task_owner: str,
+    num_slices: int = 1,
+) -> KubernetesPodOperator:
+  """Run workload through xpk tool.
+
+  The reason to use KubernetesPodOperator instead of BashOperator is that
+  xpk must run with Python 3.10 or greater; however, the latest version in
+  Composer is Python 3.8, and it's non-trivial to upgrade it as the  Composer
```

Comment:
:facepalm: I wish they would update this

Airflow actually provides a neat tool for this situation. You can run an `@task` directly in a Docker container with a different Python environment: https://airflow.apache.org/docs/apache-airflow/stable/tutorial/taskflow.html#dependency-separation-using-docker-operator

This gives us at least one less level of indirection: a docker container on the current host, rather than a docker container on another host. There's also a corresponding `@task.kubernetes` in case we do need a cluster!

Reply:
Discussed offline. It seems SubprocessHook is not compatible with @task.docker and @task.kubernetes. Not sure if this is Composer specific.